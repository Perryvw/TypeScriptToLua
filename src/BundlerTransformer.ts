import { TranspiledFile } from "./Transpile";
import { LuaProjectTransformer } from "./CustomLuaTransformers";
import { createBlock, Statement } from "./LuaAST";

type BundlerTransformerFactory = (outFile: string) => LuaProjectTransformer;

function notUndefined<T>(value: T | undefined): value is T {
    return value !== undefined;
}

// TODO: Implement correctly, this is just a PoC!
export const bundlerTransformer: BundlerTransformerFactory = (outFile: string) => (projectFiles: TranspiledFile[]) => {
    const bundledFile: TranspiledFile = {
        fileName: outFile,
        declaration: projectFiles
            .map(f => f.declaration)
            .filter(notUndefined)
            .join("\n"),
        luaAst: createBlock(
            ([] as Statement[]).concat(
                ...projectFiles
                    .map(f => f.luaAst)
                    .filter(notUndefined)
                    .map(ast => ast.statements)
            )
        ),

    };
    return [bundledFile];
};
