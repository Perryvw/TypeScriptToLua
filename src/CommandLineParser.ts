import * as path from "path";
import * as ts from "typescript";
import { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
import * as diagnostics from "./diagnostics";

export interface ParsedCommandLine extends ts.ParsedCommandLine {
    options: CompilerOptions;
}

interface CommandLineOptionBase {
    describe: string;
    aliases?: string[];
}

interface CommandLineOptionOfEnum extends CommandLineOptionBase {
    type: "enum";
    choices: string[];
}

interface CommandLineOptionOfBoolean extends CommandLineOptionBase {
    type: "boolean";
}

type CommandLineOption = CommandLineOptionOfEnum | CommandLineOptionOfBoolean;
const optionDeclarations: Record<string, CommandLineOption> = {
    luaLibImport: {
        describe: "Specifies how js standard features missing in lua are imported.",
        type: "enum",
        choices: Object.values(LuaLibImportKind),
    },
    luaTarget: {
        aliases: ["lt"],
        describe: "Specify Lua target version.",
        type: "enum",
        choices: Object.values(LuaTarget),
    },
    noHeader: {
        describe: "Specify if a header will be added to compiled files.",
        type: "boolean",
    },
    noHoisting: {
        describe: "Disables hoisting.",
        type: "boolean",
    },
    sourceMapTraceback: {
        describe: "Applies the source map to show source TS files and lines in error tracebacks.",
        type: "boolean",
    },
};

export const version = `Version ${require("../package.json").version}`;

const helpString =
    "Syntax:   tstl [options] [files...]\n\n" +

    "Examples: tstl path/to/file.ts [...]\n" +
    "          tstl -p path/to/tsconfig.json\n\n" +

    "In addition to the options listed below you can also pass options\n" +
    "for the typescript compiler (For a list of options use tsc -h).\n" +
    "Some tsc options might have no effect.";

export function getHelpString(): string {
    let result = helpString + "\n\n";

    result += "Options:\n";
    for (const [optionName, option] of Object.entries(optionDeclarations)) {
        const aliasStrings = (option.aliases || []).map(a => "-" + a);

        const optionString = aliasStrings.concat(["--" + optionName]).join("|");

        const parameterDescribe = option.type === "enum" ? option.choices.join("|") : option.type;

        const spacing = " ".repeat(Math.max(1, 45 - optionString.length - parameterDescribe.length));

        result += `\n ${optionString} <${parameterDescribe}>${spacing}${option.describe}\n`;
    }

    return result;
}

export function updateParsedConfigFile(parsedConfigFile: ts.ParsedCommandLine): ParsedCommandLine {
    for (const key in parsedConfigFile.raw) {
        const option = optionDeclarations[key];
        if (!option) continue;

        const value = readValue(parsedConfigFile.raw[key], option.type);
        if (option.type === "enum" && !option.choices.includes(value as string)) {
            parsedConfigFile.errors.push(
                diagnostics.argumentForOptionMustBe(key, option.choices.join(", "))
            );
        } else {
            // console.warn(`[Deprectated] TSTL options are moving to the luaConfig object. Adjust your tsconfig to `
            //    + `look like { "compilerOptions": { <typescript options> }, "tstl": { <tstl options> } }`);
            if (parsedConfigFile.options[key] === undefined) {
                parsedConfigFile.options[key] = value;
            }
        }
    }

    // Eventually we will only look for the tstl object for tstl options
    if (parsedConfigFile.raw.tstl) {
        for (const key in parsedConfigFile.raw.tstl) {
            const option = optionDeclarations[key];
            if (!option) continue;

            const value = readValue(parsedConfigFile.raw.tstl[key], option.type);
            if (option.type === "enum" && !option.choices.includes(value as string)) {
                parsedConfigFile.errors.push(
                    diagnostics.argumentForOptionMustBe(key, option.choices.join(", "))
                );
            } else {
                if (parsedConfigFile.options[key] === undefined) {
                    parsedConfigFile.options[key] = value;
                }
            }
        }
    }

    return parsedConfigFile;
}

export function parseCommandLine(args: string[]): ParsedCommandLine {
    const commandLine = updateParsedCommandLine(ts.parseCommandLine(args), args);

    // TODO: Remove
    if (commandLine.options.project && !commandLine.options.rootDir) {
        commandLine.options.rootDir = path.dirname(commandLine.options.project);
    }

    if (!commandLine.options.rootDir) {
        commandLine.options.rootDir = process.cwd();
    }

    if (!commandLine.options.outDir) {
        commandLine.options.outDir = commandLine.options.rootDir;
    }

    return commandLine;
}

function updateParsedCommandLine(
    parsedCommandLine: ts.ParsedCommandLine,
    args: string[]
): ParsedCommandLine {
    // Generate a list of valid option names and aliases
    const optionNames = Object.keys(optionDeclarations)
        .map(n => `--${n}`)
        .concat(...Object.values(optionDeclarations).map(o => (o.aliases || []).map(a => `-${a}`)));

    // Ignore errors caused by tstl specific compiler options
    const tsInvalidCompilerOptionErrorCode = 5023;
    parsedCommandLine.errors = parsedCommandLine.errors.filter(err => {
        return !(
            err.code === tsInvalidCompilerOptionErrorCode &&
            optionNames.some(optionName => String(err.messageText).endsWith(`'${optionName}'.`))
        );
    });

    for (let i = 0; i < args.length; i++) {
        if (!args[i].startsWith("-")) continue;

        const hasTwoDashes = args[i].startsWith("--");
        const parameterValue = args[i].substr(hasTwoDashes ? 2 : 1);
        let argumentName = optionDeclarations[parameterValue] && parameterValue;
        if (!hasTwoDashes && argumentName === undefined) {
            for (const key in optionDeclarations) {
                if ((optionDeclarations[key].aliases || []).includes(parameterValue)) {
                    argumentName = key;
                    break;
                }
            }
        }

        if (argumentName !== undefined) {
            const argumentResult = getArgumentValue(argumentName, i, args);
            if (argumentResult.isValid === true) {
                parsedCommandLine.options[argumentName] = argumentResult.result;
                // Skip value from being considered as option
                i += argumentResult.increment;
            } else {
                parsedCommandLine.errors.push(argumentResult.error);
            }
        }
    }

    return parsedCommandLine;
}

type ArgumentParseResult =
    | { isValid: true; result: string | boolean; increment: number }
    | { isValid: false; error: ts.Diagnostic };

function getArgumentValue(
    argumentName: string,
    argumentIndex: number,
    args: string[]
): ArgumentParseResult {
    const option = optionDeclarations[argumentName];
    const argument = args[argumentIndex + 1];

    if (option.type === "boolean" && (argument === undefined || argument.startsWith("-"))) {
        // Set boolean arguments without supplied value to true
        return { isValid: true, result: true, increment: 0 };
    }

    if (argument === undefined) {
        return { isValid: false, error: diagnostics.compilerOptionExpectsAnArgument(argumentName) };
    }

    const value = readValue(argument, option.type);

    if (option.type === "enum" && option.choices && !option.choices.includes(value as string)) {
        return {
            isValid: false,
            error: diagnostics.argumentForOptionMustBe(
                `--${argumentName}`,
                option.choices.join(", ")
            ),
        };
    }

    return { isValid: true, result: value, increment: 1 };
}

function readValue(value: string | boolean, type: CommandLineOption["type"]): string | boolean {
    if (type === "boolean") {
        return value === true || value === "true" || value === "t";
    } else if (type === "enum") {
        return value.toString().toLowerCase();
    }
}

export function parseConfigFileWithSystem(
    configFileName: string,
    commandLineOptions: CompilerOptions,
    system = ts.sys
): ParsedCommandLine {
    const { config, error } = ts.readConfigFile(configFileName, system.readFile);
    if (error) return { options: {}, fileNames: [], errors: [error] };

    const parsedConfigFile = ts.parseJsonConfigFileContent(
        config,
        system,
        path.dirname(configFileName),
        commandLineOptions,
        configFileName
    );

    return updateParsedConfigFile(parsedConfigFile);
}
