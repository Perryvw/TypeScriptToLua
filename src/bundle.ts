import * as path from "path";
import { SourceNode } from "source-map";
import * as ts from "typescript";
import { couldNotFindBundleEntryPoint } from "./diagnostics";
import { EmitHost, TranspiledFile } from "./Transpile";
import { formatPathToLuaPath, trimExtension } from "./utils";
import { escapeString } from "./TSHelper";
import { CompilerOptions } from "./CompilerOptions";

const formatPath = (path: string) => escapeString(formatPathToLuaPath(trimExtension(path)));
const modulePath = (baseDir: string, pathToResolve: string) => formatPath(path.relative(baseDir, pathToResolve));

export function bundleTranspiledFiles(
    bundleFile: string,
    entryModule: string,
    transpiledFiles: TranspiledFile[],
    program: ts.Program,
    emitHost: EmitHost
): [ts.Diagnostic[], TranspiledFile] {
    const diagnostics: ts.Diagnostic[] = [];

    const options = program.getCompilerOptions() as CompilerOptions;

    const projectRootDir = options.configFilePath
        ? path.dirname(options.configFilePath)
        : emitHost.getCurrentDirectory();

    // Resolve project settings relative to project file.
    const resolvedEntryModule = path.resolve(projectRootDir, entryModule);
    const resolvedBundleFile = path.resolve(projectRootDir, bundleFile);

    // Resolve source files relative to common source directory.
    const sourceRootDir = program.getCommonSourceDirectory();
    if (!transpiledFiles.some(f => path.resolve(sourceRootDir, f.fileName) === resolvedEntryModule)) {
        return [[couldNotFindBundleEntryPoint(entryModule)], { fileName: bundleFile }];
    }

    // For each file: ["<module path>"] = function() <lua content> end,
    const moduleTableEntries: SourceChunk[] = transpiledFiles.map(f =>
        moduleSourceNode(f, modulePath(sourceRootDir, f.fileName))
    );

    // If any of the modules contains a require for lualib_bundle, add it to the module table.
    const lualibRequired = transpiledFiles.some(f => f.lua && f.lua.includes(`require("lualib_bundle")`));
    if (lualibRequired) {
        const lualibBundle = emitHost.readFile(path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua"));
        moduleTableEntries.push(`["lualib_bundle"] = function() ${lualibBundle} end,\n`);
    }

    // Create ____modules table containing all entries from moduleTableEntries
    const moduleTable = createModuleTableNode(moduleTableEntries);

    // Override `require` to read from ____modules table.
    const requireOverride = `
local ____moduleCache = {}
local ____originalRequire = require
function require(file)
    if ____moduleCache[file] then return ____moduleCache[file] end
    if ____modules[file] then
        ____moduleCache[file] = ____modules[file]()
        return ____moduleCache[file]
    else
        print("Could not find module '"..file.."' to require.")
        return ____originalRequire(file) 
    end
end\n`;

    // return require("<entry module path>")
    const entryPoint = `return require("${modulePath(sourceRootDir, resolvedEntryModule)}")\n`;

    const bundleNode = joinSourceChunks([moduleTable, requireOverride, entryPoint]);
    const { code, map } = bundleNode.toStringWithSourceMap();

    return [
        diagnostics,
        {
            fileName: resolvedBundleFile,
            lua: code,
            sourceMap: map.toString(),
            sourceMapNode: moduleTable,
            declaration: undefined,
            declarationMap: undefined,
        },
    ];
}

function moduleSourceNode(transpiledFile: TranspiledFile, modulePath: string): SourceNode {
    const tableEntryHead = `["${modulePath}"] = function() `;
    const tableEntryTail = `end,\n`;

    if (transpiledFile.lua && transpiledFile.sourceMapNode) {
        return joinSourceChunks([tableEntryHead, transpiledFile.sourceMapNode, tableEntryTail]);
    } else {
        return joinSourceChunks([tableEntryHead, tableEntryTail]);
    }
}

function createModuleTableNode(fileChunks: SourceChunk[]): SourceNode {
    const tableHead = `local ____modules = {\n`;
    const tableEnd = `}\n`;

    return joinSourceChunks([tableHead, ...fileChunks, tableEnd]);
}

type SourceChunk = string | SourceNode;
function joinSourceChunks(chunks: SourceChunk[]): SourceNode {
    // tslint:disable-next-line:no-null-keyword
    return new SourceNode(null, null, null, chunks);
}
