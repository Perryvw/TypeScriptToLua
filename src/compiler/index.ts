export { Chunk } from "./chunk";
export * from "./compiler";
export { Module } from "./module";
export { Plugin } from "./plugins";
export * from "./transpile";
