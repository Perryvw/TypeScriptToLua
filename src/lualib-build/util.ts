import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
