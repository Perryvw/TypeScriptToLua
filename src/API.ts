import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileContent } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspiledFile, TranspilationResult } from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(
    fileName: string,
    options?: CompilerOptions
): TranspilationResult {
    const parseResult = parseConfigFileContent(
        fs.readFileSync(fileName, "utf8"),
        fileName,
        options
    );
    if (parseResult.isValid === false) {
        // TODO: Return diagnostics
        throw new Error(parseResult.errorMessage);
    }

    return transpileFiles(parseResult.result.fileNames, parseResult.result.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface TranspileStringResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    input: string | Record<string, string>,
    options: CompilerOptions = {}
): TranspileStringResult {
    const programFiles = typeof input === "object" ? input : { "main.ts": input };
    const mainFileName =
        typeof input === "string"
            ? "main.ts"
            : Object.keys(input).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) throw new Error('Input should have a file named "main"');

    const program = createVirtualProgram(programFiles, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
