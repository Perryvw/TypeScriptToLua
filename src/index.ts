import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspilationResult, TranspiledFile } from "./Transpile";

export { parseCommandLine, ParsedCommandLine, updateParsedConfigFile } from "./CommandLineParser";
export { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
export * from "./Emit";
export * from "./LuaAST";
export { LuaLibFeature } from "./LuaLib";
export { LuaPrinter } from "./LuaPrinter";
export { LuaTransformer } from "./LuaTransformer";
export * from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(fileName: string, options?: CompilerOptions): TranspilationResult {
    const parseResult = parseConfigFileWithSystem(fileName, options);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, transpiledFiles: new Map() };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface VirtualProgramResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    content: string,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const program = createVirtualProgram({ "main.ts": content }, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get("main.ts"), diagnostics: [...allDiagnostics] };
}

export function transpileVirtualProgram(
    files: Record<string, string>,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const mainFileName = Object.keys(files).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) {
        throw new Error('Virtual program should have a file named "main"');
    }

    const program = createVirtualProgram(files, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
