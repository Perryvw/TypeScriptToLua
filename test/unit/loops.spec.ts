import { Expect, Test, TestCase } from "alsatian";
import { LuaTarget } from "../../src/Transpiler";
import * as util from "../src/util";

const deepEqual = require("deep-equal");

export class LuaLoopTests {

    @TestCase([0, 1, 2, 3], [1, 2, 3, 4])
    @Test("while")
    public while(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            let i = 0;
            while (i < arrTest.length) {
                arrTest[i] = arrTest[i] + 1;
                i++;
            }
            return JSONStringify(arrTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3, 4], [0, 1, 2, 1, 4])
    @Test("while with continue")
    public whileWithContinue(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            let i = 0;
            while (i < arrTest.length) {
                if (i % 2 == 0) {
                    i++;
                    continue;
                }
                let j = 2;
                while (j > 0) {
                    if (j == 2) {
                        j--
                        continue;
                    }
                    arrTest[i] = j;
                    j--;
                }

                i++;
            }
            return JSONStringify(arrTest);`
        );

        // Executre
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3, 4], [0, 1, 2, 1, 4])
    @Test("dowhile with continue")
    public dowhileWithContinue(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            let i = 0;
            do {
                if (i % 2 == 0) {
                    i++;
                    continue;
                }
                let j = 2;
                do {
                    if (j == 2) {
                        j--
                        continue;
                    }
                    arrTest[i] = j;
                    j--;
                } while (j > 0)

                i++;
            } while (i < arrTest.length)
            return JSONStringify(arrTest);`
        );

        // Executre
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3], [1, 2, 3, 4])
    @Test("for")
    public for(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            for (let i = 0; i < arrTest.length; ++i) {
                arrTest[i] = arrTest[i] + 1;
            }
            return JSONStringify(arrTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3, 4], [0, 0, 2, 0, 4])
    @Test("for with continue")
    public forWithContinue(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            for (let i = 0; i < arrTest.length; i++) {
                if (i % 2 == 0) {
                    continue;
                }

                for (let j = 0; j < 2; j++) {
                    if (j == 1) {
                        continue;
                    }
                    arrTest[i] = j;
                }
            }
            return JSONStringify(arrTest);
            `
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3], [1, 2, 3, 4])
    @Test("forMirror")
    public forMirror(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            for (let i = 0; arrTest.length > i; i++) {
                arrTest[i] = arrTest[i] + 1;
            }
            return JSONStringify(arrTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3], [0, 1, 2, 3])
    @Test("forBreak")
    public forBreak(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            for (let i = 0; i < arrTest.length; ++i) {
                break;
                arrTest[i] = arrTest[i] + 1;
            }
            return JSONStringify(arrTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3], [1, 2, 3, 4], "let i = 0; i < arrTest.length; i++")
    @TestCase([0, 1, 2, 3], [1, 2, 3, 4], "let i = 0; i <= arrTest.length - 1; i++")
    @TestCase([0, 1, 2, 3], [1, 2, 3, 4], "let i = 0; arrTest.length > i; i++")
    @TestCase([0, 1, 2, 3], [1, 2, 3, 4], "let i = 0; arrTest.length - 1 >= i; i++")
    @TestCase([0, 1, 2, 3], [1, 1, 3, 3], "let i = 0; i < arrTest.length; i += 2")
    @TestCase([0, 1, 2, 3], [1, 2, 3, 4 ], "let i = arrTest.length - 1; i >= 0; i--")
    @TestCase([0, 1, 2, 3], [0, 2, 2, 4], "let i = arrTest.length - 1; i >= 0; i -= 2")
    @TestCase([0, 1, 2, 3], [0, 2, 2, 4], "let i = arrTest.length - 1; i > 0; i -= 2")
    @Test("forheader")
    public forheader(inp: number[], expected: number[], header: string): void {
        // Transpile
        const lua = util.transpileString(
            `let arrTest = ${JSON.stringify(inp)};
            for (${header}) {
                arrTest[i] = arrTest[i] + 1;
            }
            return JSONStringify(arrTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase({ ["test1"]: 0, ["test2"]: 1, ["test3"]: 2 }, { ["test1"]: 1, ["test2"]: 2, ["test3"]: 3 })
    @Test("forin[Object]")
    public forinObject(inp: any, expected: any): void {
        // Transpile
        const lua = util.transpileString(
            `let objTest = ${JSON.stringify(inp)};
            for (let key in objTest) {
                objTest[key] = objTest[key] + 1;
            }
            return JSONStringify(objTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(deepEqual(JSON.parse(result), expected)).toBe(true);
    }

    @TestCase([1, 2, 3])
    @Test("forin[Array]")
    public forinArray(inp: number[]): void {
        // Transpile & Assert
        Expect(() =>
            util.transpileString(
                `let arrTest = ${JSON.stringify(inp)};
                for (let key in arrTest) {
                    arrTest[key]++;
                }`
            )
        ).toThrowError(Error, "Iterating over arrays with 'for in' is not allowed.");
    }

    @TestCase({a: 0, b: 1, c: 2, d: 3, e: 4}, {a: 0, b: 0, c: 2, d: 0, e: 4})
    @Test("forin with continue")
    public forinWithContinue(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let obj = ${JSON.stringify(inp)};
            for (let i in obj) {
                if (obj[i] % 2 == 0) {
                    continue;
                }

                obj[i] = 0;
            }
            return JSONStringify(obj);
            `
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2], [1, 2, 3])
    @Test("forof")
    public forof(inp: any, expected: any): void {
        // Transpile
        const lua = util.transpileString(
            `let objTest = ${JSON.stringify(inp)};
            let arrResultTest = [];
            for (let value of objTest) {
                arrResultTest.push(value + 1)
            }
            return JSONStringify(arrResultTest);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase([0, 1, 2, 3, 4], [0, 0, 2, 0, 4])
    @Test("forof with continue")
    public forofWithContinue(inp: number[], expected: number[]): void {
        // Transpile
        const lua = util.transpileString(
            `let testArr = ${JSON.stringify(inp)};
            let a = 0;
            for (let i of testArr) {
                if (i % 2 == 0) {
                    a++;
                    continue;
                }

                for (let j of [0, 1]) {
                    if (j == 1) {
                        continue;
                    }
                    testArr[a] = j;
                }
                a++;
            }
            return JSONStringify(testArr);`
        );

        // Execute
        const result = util.executeLua(lua);

        // Assert
        Expect(result).toBe(JSON.stringify(expected));
    }

    @TestCase("while (a < b) { i++; }")
    @TestCase("do { i++; } while (a < b)")
    @TestCase("for (let i = 0; i < 3; i++) {}")
    @TestCase("for (let a in b) {}")
    @TestCase("for (let a of b) {}")
    @Test("loop versions")
    public whileVersions(loop: string): void {
        // Transpile
        const lua51 = util.transpileString(loop, { luaTarget: LuaTarget.Lua51 });
        const lua52 = util.transpileString(loop, { luaTarget: LuaTarget.Lua52 });
        const lua53 = util.transpileString(loop, { luaTarget: LuaTarget.Lua53 });
        const luajit = util.transpileString(loop, { luaTarget: LuaTarget.LuaJIT });

        // Assert
        Expect(lua51.indexOf("::__continue0::") !== -1).toBe(false); // No labels in 5.1
        Expect(lua52.indexOf("::__continue0::") !== -1).toBe(true); // Labels from 5.2 onwards
        Expect(lua53.indexOf("::__continue0::") !== -1).toBe(true);
        Expect(luajit.indexOf("::__continue0::") !== -1).toBe(true);
    }

    @Test("for dead code after return")
    public forDeadCodeAfterReturn(): void {
        const result = util.transpileAndExecute(
            `for (let i = 0; i < 10; i++) { return 3; const b = 8; }`);

        Expect(result).toBe(3);
    }

    @Test("for..in dead code after return")
    public forInDeadCodeAfterReturn(): void {
        const result = util.transpileAndExecute(
            `for (let a in {"a": 5, "b": 8}) { return 3; const b = 8; }`);

        Expect(result).toBe(3);
    }

    @Test("for..of dead code after return")
    public forOfDeadCodeAfterReturn(): void {
        const result = util.transpileAndExecute(
            `for (let a of [1,2,4]) { return 3; const b = 8; }`);

        Expect(result).toBe(3);
    }

    @Test("while dead code after return")
    public whileDeadCodeAfterReturn(): void {
        const result = util.transpileAndExecute(
            `while (true) { return 3; const b = 8; }`);

        Expect(result).toBe(3);
    }
}
