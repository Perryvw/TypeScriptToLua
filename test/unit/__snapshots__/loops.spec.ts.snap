// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forin[Array] 1`] = `"main.ts(3,9): error TSTL: Iterating over arrays with 'for ... in' is not allowed."`;
